### PR TITLE
Search by chef or meals

### PR DESCRIPTION
**Pull request by gary-bains**

### CODE REVIEW

**Comment by ThinkTink** [comment]
Quick request @garys-io , can you handle the merge conflicts?

**Comment by gary-bains** [comment]
> Quick request @garys-io , can you handle the merge conflicts?

Hi, sure, I'll do that but I'm waiting on Megan to make a PR which'll resolve many of the merge issue as I have merged their branch into mine earlier.

File: client/src/components/ChefCard.js [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+    const {
+        firstName,
+        lastName,
+        bio,
+        primaryAddress: { city, region, country },
+    } = userId;
+    const location = [city, region, country]
+        .filter((ele) => ele && typeof ele === "string" && ele.length > 0)
+        .join(", ");
+    console.log(cuisineSpecialty, distanceKm, location, bio);
```
Comment (ThinkTink) [inline client/src/components/ChefCard.js]:
```suggestion
```

File: client/src/pages/EditProfile.js [hunk 2]
```diff
@@ -93,23 +93,26 @@ export default function EditProfile() {
...
-                    city: userData.primaryAddress["city"],
-                    region: userData.primaryAddress["region"],
-                    country: userData.primaryAddress["country"],
+                    street: userData.primaryAddress["street"] || "",
+                    city: userData.primaryAddress["city"] || "",
+                    region: userData.primaryAddress["region"] || "",
+                    country: userData.primaryAddress["country"] || "",
                 }}
                 onSubmit={(values, { setSubmitting }) => {
+                    console.log(values);
```
Comment (ThinkTink) [inline client/src/pages/EditProfile.js]:
```suggestion
```

File: client/src/pages/Meals.js [hunk 3]
```diff
@@ -1,35 +1,94 @@
...
+
     return (
-        <React.Fragment>
+        <Main>
             <ResponsiveSideBar>
-                <Typography>Responsive Side Bar</Typography>
+                <Grid container spacing={2} alignContent="flex-start">
+                    <Grid item xs={12}>
+                        <Typography>Cuisine:</Typography>
+                        {[...cuisines].map((cuisine) => (
```
Comment (ThinkTink) [inline client/src/pages/Meals.js:48]: Do we need `[...cuisines]` everywhere instead of just `cuisines`?

File: client/src/pages/Meals.js [hunk 4]
```diff
@@ -1,35 +1,94 @@
...
+                        key={meal._id}
+                        id={meal._id}
+                        mealPic={meal.mealPic}
+                        title={meal.title}
+                        price={meal.price}
+                        chefName={meal.chefName}
+                        chefId={meal.chefId}
+                        chefPic={meal.chefPic}
+                        location={meal.location}
+                    />
```
Comment (ThinkTink) [inline client/src/pages/Meals.js]: Is there a reason to pass each of these individuall vs just passing meal itself?

File: client/src/utils.js [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+
+    return {
+        abort: () => controller.abort(),
+        ready: fetch(request, { ...opts, signal }),
+    };
+};
+
+// If we know it's been canceled! do nothing
+const consoleErrorNonAbortErrors = (e) =>
+    e.name !== "AbortError" && console.error(e);
```
Comment (ThinkTink) [inline client/src/utils.js]: While it's fine for testing, I don't think we should plan on console logging errors. Depending on how we want to handle it, I think we should either show an error at the component level, or just ignore errors of this nature

File: server/routes/search.js [hunk 6]
```diff
@@ -1,52 +1,78 @@
...
+                _id: { $in: meals.map((meal) => meal.chefId) },
+            });
 
-    }
-    
-    return;
-});
+            const chefIndex = {};
+            chefs.forEach((chef) => (chefIndex[chef._id] = chef));
+            meals.forEach((meal) => (meal["chefId"] = chefIndex[meal.chefId]));
```
Comment (ThinkTink) [inline server/routes/search.js]: What is the purpose of this step?

File: server/routes/search.js [hunk 7]
```diff
@@ -1,52 +1,78 @@
...
-        meals.forEach(meal => meal["chefId"]= chefIndex[meal.chefId])
 
-        res.json({ meals  });
+        if (req.query.searchType === "meals") {
+            const query = req.query.cuisine
+                ? { cuisineType: { $in: req.query.cuisine.split(",").map(c => RegExp(c, 'i')) } }
+                : {};
+            const meals = await mealsController.findAllMeals(query);
+            if (!query) {
+                res.json(chefs);
```
Comment (ThinkTink) [inline server/routes/search.js]: Should this be `res.json(meals)`?

File: server/routes/search.js [hunk 8]
```diff
@@ -1,52 +1,78 @@
...
-    
-    return;
-});
+            const chefIndex = {};
+            chefs.forEach((chef) => (chefIndex[chef._id] = chef));
+            meals.forEach((meal) => (meal["chefId"] = chefIndex[meal.chefId]));
+            res.json({ meals });
+        }
+    })
+);
```
Comment (ThinkTink) [inline server/routes/search.js:49]: This whole route is a bit hard to follow. I'm thinking it might be easier to read if you pull some of the work out into functions?

File: client/src/pages/ChefSearch.js [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+    } = useCuisineSelector();
+    const { profile } = useContext(UserContext);
+    const {
+        primaryAddress: { city, region, country },
+    } = profile;
+    const location = [city, region, country].join(", ");
+    const [radiusKm, setRadiusKm] = useState(100);
+    const [results, setResults] = useState({
+        chefs: [],
+        chefMeals: {},
```
Comment (ThinkTink) [inline client/src/pages/ChefSearch.js]: Are we using this anywhere?

**Review by ThinkTink — COMMENTED** [review]
Only needs a little bit of clean up before merging, but I do want to wait to merge until Megan's upcoming PR is merged so we can make sure we don't merge anything we don't intend to

File: client/src/pages/Meals.js (hunk 3 above)
Comment (gary-bains) [reply client/src/pages/Meals.js:48]: Yes, we need it to convert the set into a list

File: client/src/pages/Meals.js (hunk 4 above)
Comment (gary-bains) [reply client/src/pages/Meals.js]: fixed it!

File: client/src/pages/ChefSearch.js (hunk 9 above)
Comment (gary-bains) [reply client/src/pages/ChefSearch.js]: refactored it to just use chefs array and removed the rest of the object

File: client/src/utils.js (hunk 5 above)
Comment (gary-bains) [reply client/src/utils.js]: Should I change it to `const consoleErrorNonAbortErrors = () => {}`

Comment (ThinkTink) [reply client/src/utils.js]: I think we can just remove that function altogether. It's just being used inside the try-catch, so we just leave the catch block empty except for a comment explaining that we don't want to notify the user of any errors there

Comment (gary-bains) [reply client/src/utils.js]: Fixed it!

File: server/routes/search.js (hunk 8 above)
Comment (gary-bains) [reply server/routes/search.js:49]: I agree, I refactored it

Comment (ThinkTink) [reply server/routes/search.js:49]: This is much better!

**Review by ThinkTink — APPROVED** [review]
Thanks for the fixes! Just make sure you look through it after Megan's PR merges to make sure there's nothing left here that we don't want to keep.
